x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java

```diff
@@ -200,14 +200,34 @@ public void addListener(ActionListener<Long> listener, long value) {
         assert invariant();
     }
 
-    private static void executeListener(final ActionListener<Long> listener, final Supplier<Long> result) {
+    /**
+     * Return the result of this future, if it has been completed successfully, or unwrap and throw the exception with which it was
+     * completed exceptionally. It is not valid to call this method if the future is incomplete.
+     */
+    private Long actionResult() throws Exception {
+        try {
+            return result();
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof Exception exCause) {
+                throw exCause;
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    private static void executeListener(final ActionListener<Long> listener, final ThrowingSupplier<Long> result) {
         try {
             listener.onResponse(result.get());
         } catch (Exception e) {
             listener.onFailure(e);
         }
     }
 
+    private interface ThrowingSupplier<T> {
+        T get() throws Exception;
+    }
```

Comment:
Duplicate of the one in `ESTestCase`, local again because we're not aiming for any grand new patterns. Perhaps there's already something we can use?